src/test/java/org/broadinstitute/hellbender/engine/ReadsDataSourceUnitTest.java

```diff
@@ -446,4 +446,87 @@ public void testQueryUnmapped( final File samFile, final List<String> expectedRe
             }
         }
     }
+
+    @DataProvider(name = "MergedHeaderIntervalQueries")
+    public Object[][] mergedHeaderQueries() {
+        return new Object[][] {
+
+                // Single interval that only overlpas one of the input sources
+                { new SimpleInterval[] {new SimpleInterval("EXTRA_CONTIG_1", 1,10)},
+                         new String[] {"EXTRA_CONTIG_1_READ"}, new int[] {4} },
+                { new SimpleInterval[] {new SimpleInterval("EXTRA_CONTIG_2", 1,10)},
+                         new String[] {"EXTRA_CONTIG_2_READ"}, new int[] {5} },
+
+                // Multiple intervals, each of which only overlaps one of the input sources
+                { new SimpleInterval[] {new SimpleInterval("EXTRA_CONTIG_1", 1,10), new SimpleInterval("EXTRA_CONTIG_2", 1,10)},
+                         new String[] {"EXTRA_CONTIG_1_READ", "EXTRA_CONTIG_2_READ"},
+                         new int[] { 4, 5} },
+
+                // Single interval doesn't overlap ANY input source
+                { new SimpleInterval[] {new SimpleInterval("TOTALLY_FAKE_CONTIG", 1,10)},
+                         new String[] {}, new int[] {} }
+        };
+    }
+
+    @Test(dataProvider = "MergedHeaderIntervalQueries")
+    public void testMergedHeaderQuery(
+            final SimpleInterval intervals[],
+            final String expectedReadNames[],
+            final int expectedSequenceIndex[]) {
+        // create two files, each with a read referencing a sequence that is not present in the other
+        final File testFile1 = getFileWithAddedContig(FIRST_TEST_BAM, "EXTRA_CONTIG_1", "test1", ".bam");
+        final File testFile2 = getFileWithAddedContig(SECOND_TEST_BAM, "EXTRA_CONTIG_2", "test2", ".bam");
+
+        ReadsDataSource readsSource = new ReadsDataSource(Arrays.asList(testFile1, testFile2));
+
+        SAMFileHeader samHeader = readsSource.getHeader();
+        SAMSequenceDictionary sequenceDictionary = readsSource.getSequenceDictionary();
+
+        // 4 sequences in the original dictionaries, plus 1 added to each of the 2 inputs == 6
+        Assert.assertEquals(sequenceDictionary.getSequences().size(), 6);
+        Assert.assertEquals(samHeader.getSequenceDictionary(), sequenceDictionary);
+
+        readsSource.setTraversalBounds(Arrays.asList(intervals));
+        int count = 0;
+        for (final GATKRead read : readsSource) {
+            Assert.assertEquals(expectedReadNames[count], read.getName());
+            Assert.assertEquals(
+                    read.convertToSAMRecord(samHeader).getReferenceIndex(),
```

Comment:
Instead of `read.convertToSAMRecord(samHeader).getReferenceIndex()` you should call `ReadUtils.getReferenceIndex()`